# Integral and mean outputs
Two new functions, `integral` and `mean`, have been added for use in model assignments.  Suppose $f = f(t, x, k)$ is an expression involving $t$, state, parameters, and local assignments, and $a, b$ are predetermined expressions.  The syntax `integral(f, a, b)` represents the integral of $f$ over the time-interval $[a, b]$, i.e. $$\int_a^b f(t, x(t), k) \, dt,$$ while `mean(f, a, b)` is the mean of $f$ over $[a, b]$, and is equivalent to `integral(f, a, b) / (b - a)`.

Integrals are another example of *nonlocal* expressions (like indexing expressions `x[t]`); they are functions of model state over the entire given time interval.  Integrals may be combined with other expressions, albeit with certain limitations similar to those of indexed expressions.  A simple model file `decay_with_integrals.model` is included for demonstration purposes, and has more detail on syntax and limitations. 

In [2]:
import abm

In [3]:
# Load and simulate model with example integral outputs.  See model file for more details.
simulation = abm.simulate(models = "model_integral.model", times=[1, 2, 3]).to_pandas()
display(simulation)

'Total: 1, Running: 0, Succeeded: 1, Failed: 0, Stopped: 0'

Simulations finished


id    t         u   z1   z2        y1        y2        y3        y4  \
0   0  1.0  0.735759  0.3  4.4  0.864664  0.432332  1.457081  0.896771   
1   0  2.0  0.270671  0.3  4.4  0.864664  0.432332  1.457081  0.896771   
2   0  3.0  0.099575  0.3  4.4  0.864664  0.432332  1.457081  0.896771   

         y5        y6    k   t1   t2         a  
0  1.457081  0.355386  1.0  0.1  1.1  0.367879  
1  1.457081  0.355386  1.0  0.1  1.1  0.135336  
2  1.457081  0.355386  1.0  0.1  1.1  0.049787

In particular, we can see `y1` is the familiar result $$\int_0^2 e^{-t} \, dt = -\left(e^{-2} - e^{0} \right) \approx 0.86. $$